compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/ConstExpr.cpp

```diff
@@ -36,8 +36,43 @@ OpOperand *findOperandFor(Operation *op, Value input) {
 }
 
 bool isHoistableToRootOp(Operation *rootOp, Operation *constOp) {
-  Operation *syms = SymbolTable::getNearestSymbolTable(constOp);
-  return syms == rootOp;
+  if (constOp->isRegistered()) {
+    Operation *syms = SymbolTable::getNearestSymbolTable(constOp);
+    if (syms != rootOp)
+      return false;
+  } else {
+    // Special handling for unregistered testing ops.
+    // Since these are unregistered, SymbolTable::getNearestSymbolTable can't
+    // check if they define new symbol tables.
+    auto opName = constOp->getName().getStringRef();
+    if (opName != "iree_unregistered.var_expr" &&
+        opName != "iree_unregistered.non_leaf_const_expr" &&
+        opName != "iree_unregistered.const_expr") {
```

Comment:
(these have always made me sad :( we should find a way to remove these in the future)

Reply:
Yeah, it takes a big chunk of code... Maybe define a test dialect or move tests `hoist_into_globals_linalg.mlir` and rewrite with linalg ops?

Reply:
we could base it on attrs instead - then any op can be used and this can just check for the attr - this would allow overrides by users as well

Reply:
I remembered Quinn was working on an `HoistableOpInterface` to mark if an op is hoistable: #15714

@qedawkins What do you think about adding an attribute to mark if op is hoistable?

Reply:
+1 to attributes, and I don't think the hoisting interface and the attributes are mutually exclusive. Two passes sounds a bit extreme to me also. We could just let attributes take precedence over any interface (if we do decide to end up landing the op interface).

Reply:
yeah - can just treat the attr here as an OR with the interface - that way it can be added to out of tree ops and such too

Reply:
+1 for the attribute - definitely for testing. Should also be fine for end user in most cases  (I wan't to be sure we can support what the user wants and with plugins etc we cannot provide as many guarantees ...)